File: ts/src/frontDoor/render/journeys.ts
// `journeys` rendering: read journey options, enrich with stage/description,
// format the roots-then-children hierarchy. Extracted from cli.ts (CR002).

import type { Database } from "../../db/database.ts";
import { optionalString, requireString } from "../../db/rowDecode.ts";
import {
  groupJourneysByParent,
  type JourneyIdentityRow,
  type JourneyOption,
  listJourneyOptions,
} from "../../journey/journeyOptions.ts";
import { identityRows } from "./identityRows.ts";

/** A journey option enriched with the display-only stage and description. */
export interface JourneyDisplayRow extends JourneyOption {
  stage: string;
  description: string;
}

const STATUS_ICONS: Record<string, string> = { active: "🚧", completed: "✅", paused: "⏸" };

/** Read the `content` column of an identity layer keyed by identity key. */
function contentByKey(db: Database, layer: string): Map<string, string> {
  return new Map(
    identityRows(db, layer).map((row) => [
      requireString(row, "key"),
      optionalString(row, "content") ?? "",
    ]),
  );
}

/**
 * Read journeys and enrich each with its stage (from journey_path) and
 * description. Two queries total (journey + journey_path layers), not the
 * former 1 + 2N: the journey content is loaded once into a map and the stage
 * layer once, instead of a per-journey pair of lookups.
 */
export function journeyRows(db: Database): JourneyDisplayRow[] {
  const journeyContent = contentByKey(db, "journey");
  const stageContent = contentByKey(db, "journey_path");
  const identityForOptions: JourneyIdentityRow[] = identityRows(db, "journey").map((row) => ({
    key: requireString(row, "key"),
    content: optionalString(row, "content") ?? "",
    metadata: optionalString(row, "metadata"),
  }));
  return listJourneyOptions(identityForOptions).map((option) => {
    const desc = (journeyContent.get(option.id) ?? "")
      .split("\n")
      .map((line) => line.trim())
      .find((line) => line && !line.startsWith("#") && !line.startsWith("**"));
    const stage =
      (stageContent.get(option.id) ?? "")
        .match(/\*\*(?:Current stage|Etapa atual):\*\*\s*(.+)/)?.[1]
        ?.trim() ?? "—";
    return { ...option, stage, description: (desc ?? "").slice(0, 80) };
  });
}

/** Format one journey (root or child) into display lines. */
export function renderJourney(row: JourneyDisplayRow, child = false): string[] {
  const icon = STATUS_ICONS[row.status] ?? "•";
  const prefix = child ? "  └─ " : "";
  const detailIndent = child ? "       " : "  ";
  const lines = [
    `${prefix}${icon} **${row.id}** (${row.status})`,
    `${detailIndent}Stage: ${row.stage}`,
  ];
  if (row.description) lines.push(`${detailIndent}${row.description}`);
  lines.push("");
  return lines;
}

/** Render the full `journeys` listing, roots each immediately followed by children. */
export function renderJourneys(db: Database): string {
  const rows = journeyRows(db);
  if (rows.length === 0) return "No journeys found.\n";
  const { roots, childrenByParent } = groupJourneysByParent(rows);
  const lines: string[] = [];
  for (const row of roots) {
    lines.push(...renderJourney(row));
    for (const child of childrenByParent.get(row.id) ?? [])
      lines.push(...renderJourney(child, true));
  }
  return lines.join("\n");
}

File: src/memory/cli/journeys.py
"""CLI: list journeys with status, stage, and description."""

import argparse
import re

from memory import MemoryClient
from memory.cli.common import db_path_from_mirror_home


def _journey_row(
    mem: MemoryClient,
    option: dict,
    last_interactions: dict[str, str] | None = None,
) -> dict:
    name = option["id"]
    ident = mem.store.get_identity("journey", name)
    content = ident.content if ident else ""
    status = option.get("status") or "?"

    desc = ""
    for line in content.split("\n"):
        line = line.strip()
        if line.startswith(("## Description", "## Descrição")):
            continue
        if line and not line.startswith("#") and not line.startswith("**"):
            desc = line[:80]
            break

    journey_path_raw = mem.get_identity("journey_path", name)
    journey_path = journey_path_raw if isinstance(journey_path_raw, str) else ""
    stage_match = re.search(r"\*\*(?:Current stage|Etapa atual):\*\*\s*(.+)", journey_path)
    stage = stage_match.group(1).strip() if stage_match else "—"

    last_raw = (last_interactions or {}).get(name, "")
    last = last_raw[:10] if last_raw else "—"
    return {
        "id": name,
        "status": status,
        "stage": stage,
        "description": desc,
        "last_interaction": last,
        "parent_journey": option.get("parent_journey") or "",
        "depth": option.get("depth", 0),
    }


def _print_journey(row: dict) -> None:
    status = row["status"]
    icon = {"active": "🚧", "completed": "✅", "paused": "⏸"}.get(status, "•")
    depth = int(row.get("depth", 0))
    prefix = f"{'│  ' * depth}└─ " if depth else ""
    detail_indent = f"{'│  ' * depth}  " if depth else "  "
    print(f"{prefix}{icon} **{row['id']}** ({status})")
    print(f"{detail_indent}Stage: {row['stage']} · Last: {row.get('last_interaction', '—')}")
    if row["description"]:
        print(f"{detail_indent}{row['description']}")
    print()


def main(argv: list[str] | None = None) -> None:
    parser = argparse.ArgumentParser(description="List journeys with status and stage")
    parser.add_argument(
        "--mirror-home",
        default=None,
        help="Explicit user home whose database should be read for this command",
    )
    args = parser.parse_args(argv)

    mem = MemoryClient(db_path=db_path_from_mirror_home(args.mirror_home))
    options = mem.journeys.list_journey_options()

    if not options:
        print("No journeys found.")
        return

    last_interactions = mem.journeys.last_interactions()
    rows = [_journey_row(mem, option, last_interactions) for option in options]
    for row in rows:
        _print_journey(row)


if __name__ == "__main__":
    main()
